### ISSUE TITLE
[MAUI] MacCatalyst parallel build issue

### ISSUE THREAD

**Issue by jonathanpeppers** [xamarin/xamarin-macios#17497]
### Steps to Reproduce

1. Build dotnet/maui's repo, namely for MacCatalyst
2. Sometimes you get errors about `.pdb` files such as:

```
Errors
    /Users/builder/azdo/_work/2/s/bin/dotnet/sdk/7.0.102/Microsoft.Common.CurrentVersion.targets(4752,5): error MSB3027: Could not copy "obj/Release/net7.0-maccatalyst/Microsoft.Maui.pdb" to "bin/Release/net7.0-maccatalyst/Microsoft.Maui.pdb". Exceeded retry count of 10. Failed.  [/Users/builder/azdo/_work/2/s/src/Core/src/Core.csproj]
    /Users/builder/azdo/_work/2/s/bin/dotnet/sdk/7.0.102/Microsoft.Common.CurrentVersion.targets(4752,5): error MSB3021: Unable to copy file "obj/Release/net7.0-maccatalyst/Microsoft.Maui.pdb" to "bin/Release/net7.0-maccatalyst/Microsoft.Maui.pdb". The process cannot access the file '/Users/builder/azdo/_work/2/s/src/Core/src/bin/Release/net7.0-maccatalyst/Microsoft.Maui.pdb' because it is being used by another process. [/Users/builder/azdo/_work/2/s/src/Core/src/Core.csproj]
```

I think this occurs if you have multiple MacCatalyst app projects in a solution and build them. I think it might have been introduced when `RuntimeIdentifiers=maccatalyst-x64;maccatalyst-arm64` became the default? But also might be combination of some change in the .NET SDK or MSBuild.

The workaround is either:

* Build `.sln` files with `-m:1`
* Build with `-r maccatalyst-x64`

This is what I did for now:

https://github.com/dotnet/maui/pull/13201/commits/440fd194b22649b0390f15865cf6cb94c6bc6017#diff-74e588eb60cf3c891901dd5657c29e2b0b0b1e605a91d5f04220301ff4a5295bR14

### Expected Behavior

No build errors.

### Actual Behavior

Got a build error.

### Environment

This was the versions from https://github.com/dotnet/maui/pull/13201

### Build Logs

Too large for GitHub:

https://microsoft-my.sharepoint.com/:u:/p/jopepper/EY6Tj2IrVkxGiVeOjABGqCsBrSnSwAtaxE72O_Fy8RV8YQ?e=XCPawx

**Comment by rolfbjarne** [comment]
Is it always `Microsoft.Maui.pdb`, or does it happen with pdbs from other assemblies as well?

It's also interesting to note that the source file and the destination file are identical, so it's trying to copy a file over itself.

**Comment by jonathanpeppers** [comment]
I think I saw it do `Microsoft.Maui.Maps.pdb` once. I don't think I was able to reproduce it locally, though...

**Comment by jonathanpeppers** [comment]
@rolfbjarne here is another example:

https://dev.azure.com/xamarin/public/_build/results?buildId=79089&view=logs&jobId=fa417ba0-7221-55e8-0ddb-67022373081a&j=fa417ba0-7221-55e8-0ddb-67022373081a&t=28a777bb-b552-5226-8cc7-96e0eb4324c7

```
/Users/builder/azdo/_work/1/s/bin/dotnet/sdk/7.0.103/Microsoft.Common.CurrentVersion.targets(4752,5): error MSB3027: Could not copy "obj/Release/net7.0-maccatalyst/Microsoft.Maui.Controls.Compatibility.pdb" to "bin/Release/net7.0-maccatalyst/Microsoft.Maui.Controls.Compatibility.pdb". Exceeded retry count of 10. Failed.  [/Users/builder/azdo/_work/1/s/src/Compatibility/Core/src/Compatibility.csproj::TargetFramework=net7.0-maccatalyst]
/Users/builder/azdo/_work/1/s/bin/dotnet/sdk/7.0.103/Microsoft.Common.CurrentVersion.targets(4752,5): error MSB3021: Unable to copy file "obj/Release/net7.0-maccatalyst/Microsoft.Maui.Controls.Compatibility.pdb" to "bin/Release/net7.0-maccatalyst/Microsoft.Maui.Controls.Compatibility.pdb". The process cannot access the file '/Users/builder/azdo/_work/1/s/src/Compatibility/Core/src/bin/Release/net7.0-maccatalyst/Microsoft.Maui.Controls.Compatibility.pdb' because it is being used by another process. [/Users/builder/azdo/_work/1/s/src/Compatibility/Core/src/Compatibility.csproj::TargetFramework=net7.0-maccatalyst]
/Users/builder/azdo/_work/1/s/bin/dotnet/sdk/7.0.103/Microsoft.Common.CurrentVersion.targets(4752,5): error MSB3027: Could not copy "obj/Release/net7.0-maccatalyst/Microsoft.Maui.Essentials.pdb" to "bin/Release/net7.0-maccatalyst/Microsoft.Maui.Essentials.pdb". Exceeded retry count of 10. Failed.  [/Users/builder/azdo/_work/1/s/src/Essentials/src/Essentials.csproj::TargetFramework=net7.0-maccatalyst]
/Users/builder/azdo/_work/1/s/bin/dotnet/sdk/7.0.103/Microsoft.Common.CurrentVersion.targets(4752,5): error MSB3021: Unable to copy file "obj/Release/net7.0-maccatalyst/Microsoft.Maui.Essentials.pdb" to "bin/Release/net7.0-maccatalyst/Microsoft.Maui.Essentials.pdb". The process cannot access the file '/Users/builder/azdo/_work/1/s/src/Essentials/src/bin/Release/net7.0-maccatalyst/Microsoft.Maui.Essentials.pdb' because it is being used by another process. [/Users/builder/azdo/_work/1/s/src/Essentials/src/Essentials.csproj::TargetFramework=net7.0-maccatalyst]
```

You want this `.binlog`:

![image](https://user-images.githubusercontent.com/840039/218281643-4d8a3581-8ea4-4485-9fa2-281052167e2a.png)

![image](https://user-images.githubusercontent.com/840039/218281646-fde4cb7e-1dec-44bb-b77a-14f39db66c0d.png)

**Comment by rolfbjarne** [comment]
@jonathanpeppers have you seen this recently?

**Comment by mattleibow** [comment]
Yeah, I tried to remove the RID in the csproj but it just failed with all the blazor things. I started a new PR for the total removal as that PR is just to make it work today.

This PR is the remove all explicit rids: https://github.com/dotnet/maui/pull/16765

**Comment by jeremy-visionaid** [comment]
I'm seeing a now relatively frequent build collision occasionally on CI for DLLs in a large-ish MAUI project. Seems like it tries to build dependencies twice for catalyst release (maccatalyst-x64 and maccatalyst-arm64) with the same path (libraries don't have the runtime identifier included in the `OutputPath` and `AppendRuntimeIdentifierToOutputPath` only seems to apply to `RuntimeIdentifier` and not `RuntimeIdentifiers`). In Visual Studio the collision doesn't seem to occur though (and I can see that it only tries to build the catalyst DLLs once there).

**Comment by rolfbjarne** [comment]
We build multiple RuntimeIdentifiers using the MSBuild task:

https://github.com/xamarin/xamarin-macios/blob/bb965c1724ef510d818171186f5085ef87b3f71c/dotnet/targets/Xamarin.Shared.Sdk.targets#L390-L402

and the MSBuild task is by default *not* parallel: "BuildInParallel=If true, the projects specified in the Projects parameter are built in parallel if it is possible. Default is false." (https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-task?view=vs-2022)

So it's somewhat of a mystery how this happens.

@jeremy-visionaid do you have an [MSBuild binlog](https://github.com/xamarin/xamarin-macios/wiki/Diagnosis#binary-build-logs) for a build that fails?

**Comment by jeremy-visionaid** [comment]
@rolfbjarne Thanks for getting back to me so quickly! I was curious why I was getting different results on CI. Seems like it was failing with VS BuildTools 17.10.6 LTSC with MSBuild 17.10.4, but it works locally with 17.11.0. Updating the CI environment resolves the problem. I could share the binlog with you if there's perhaps something that needs backporting to LTSC, but seems that whatever the problem is already fixed upstream - I can just use the current channel.

**Comment by rolfbjarne** [comment]
> @rolfbjarne Thanks for getting back to me so quickly! I was curious why I was getting different results on CI. Seems like it was failing with VS BuildTools 17.10.6 LTSC (with MSBuild 17.10.4), but it works locally with 17.11.0. Updating the CI environment resolves the problem. I could share the binlog with you if there's perhaps something that needs backporting to LTSC, but seems that whatever the problem is already fixed upstream - I can just use the current channel.

That's weird... I'm not aware of any fixes in this area. Parallel issues can be annoying to track down though, some completely different thing might cause the build to happen in a different order, thus avoiding the problem.

If you have a binlog already that shows the error, I'd love to see it, otherwise don't worry about it.

**Comment by jeremy-visionaid** [comment]
@rolfbjarne Yeah, it's hard to be sure whether it truly resolves the problem yet. I was wrong about it colliding between arm64 and x64 - the error can also still occur on 17.10 if I only build `maccatalyst-x64`, and I can see the collision in the structured log viewer. It was occuring on most runs on 17.10 but I've yet to repro the issue on 17.11 after a few dozen runs now. I'm happy to share the binlogs with you - maybe I'd just done something wrong/silly, but is there a way I can provide it to you privately?

**Comment by rolfbjarne** [comment]
@jeremy-visionaid if you can put them on Dropbox or a similar service, and then send the link to my email (Rolf.Kvinge@microsoft.com)

**Comment by rolfbjarne** [comment]
With help from @jeremy-visionaid, I was able to determine that this was fixed with https://github.com/xamarin/xamarin-macios/pull/19990.
